### PR TITLE
Download go modules before building

### LINKED ISSUE #14: Unable to build binary due to missing go.sum
I am using `0.7.0` installed via `go get` and `Go 1.16`.

My config file is:
```yaml
exporters:
  - gomod: "github.com/open-telemetry/opentelemetry-collector-contrib/exporter/newrelicexporter v0.21.0"
```

Upon executing the builder using `opentelemetry-collector-builder --output-path=/tmp/dist --config=.otelcol-builder.yaml`, I get the following error:

```
2021-03-05T15:50:35.630+1100    INFO    cmd/root.go:82  OpenTelemetry Collector distribution builder    {"version": "dev", "date": "unknown"}
2021-03-05T15:50:35.630+1100    INFO    cmd/root.go:98  Using config file       {"path": ".otelcol-builder.yaml"}
2021-03-05T15:50:35.633+1100    INFO    builder/main.go:80      Sources created {"path": "/tmp/dist"}
2021-03-05T15:50:35.666+1100    INFO    builder/main.go:94      Using go from PATH      {"Go executable": "/usr/local/go/bin/go"}
2021-03-05T15:50:35.666+1100    INFO    builder/main.go:97      Compiling
Error: failed to compile the OpenTelemetry Collector distribution: exit status 1. Output: "go: github.com/open-telemetry/opentelemetry-collector-contrib/exporter/newrelicexporter@v0.21.0: missing go.sum entry; to add it:\n\tgo mod download github.com/open-telemetry/opentelemetry-collector-contrib/exporter/newrelicexporter\n"
Usage:
  opentelemetry-collector-builder [flags]

Flags:
      --config string            config file (default is $HOME/.otelcol-builder.yaml)
      --description string       A descriptive name for the OpenTelemetry Collector distribution (default "Custom OpenTelemetry Collector distribution")
      --go string                The Go binary to use during the compilation phase (default "/usr/bin/go")
  -h, --help                     help for opentelemetry-collector-builder
      --include-core             Whether the core components should be included in the distribution (default true)
      --module string            The Go module for the new distribution (default "github.com/jpkroehling/opentelemetry-collector-builder")
      --name string              The executable name for the OpenTelemetry Collector distribution (default "otelcol-custom")
      --otelcol-version string   Which version of OpenTelemetry Collector to use as base (default "0.20.0")
      --output-path string       Where to write the resulting files (default "/tmp/otelcol-distribution776575606")
      --version string           The version for the OpenTelemetry Collector distribution (default "1.0.0")

2021-03-05T15:50:35.702+1100    ERROR   cmd/root.go:77  failed to run   {"error": "failed to compile the OpenTelemetry Collector distribution: exit status 1. Output: \"go: github.com/open-telemetry/opentelemetry-collector-contrib/exporter/newrelicexporter@v0.21.0: missing go.sum entry; to add it:\\n\\tgo mod download github.com/open-telemetry/opentelemetry-collector-contrib/exporter/newrelicexporter\\n\""}
github.com/open-telemetry/opentelemetry-collector-builder/cmd.Execute
        /c/Users/User/go/pkg/mod/github.com/open-telemetry/opentelemetry-collector-builder@v0.7.0/cmd/root.go:77
main.main
        /c/Users/User/go/pkg/mod/github.com/open-telemetry/opentelemetry-collector-builder@v0.7.0/main.go:22
runtime.main
        /usr/local/go/src/runtime/proc.go:225
```

From my understanding, the builder should take care of initializing the `go.mod` and `go.sum` files and user intervention is not required.

### PR DESCRIPTION
**Pull request by eroteme**
Fixes #14 

Since go 1.16 build no longer  updates go.sum and so we get errors such as:

```
go: github.com/open-telemetry/opentelemetry-collector-contrib/processor/routingprocessor@v0.22.0: missing go.sum entry; to add it:\n\tgo mod download github.com/open-telemetry/opentelemetry-collector-contrib/processor/routingprocessor\n"
```

This change runs `go mod download` to update go sum before the compile

### CODE REVIEW

**Comment by eroteme** [comment]
Issue #14

File: README.md [hunk 1]
```diff
@@ -7,7 +7,7 @@ This program generates a custom OpenTelemetry Collector binary based on a given
 $ GO111MODULE=on go get github.com/open-telemetry/opentelemetry-collector-builder
 $ cat > ~/.otelcol-builder.yaml <<EOF
 exporters:
-  - gomod: "github.com/open-telemetry/opentelemetry-collector-contrib/exporter/alibabacloudlogserviceexporter v0.21.0"
+  - gomod: "github.com/open-telemetry/opentelemetry-collector-contrib/exporter/alibabacloudlogserviceexporter v0.22.0"
```
Comment (jpkrohling) [inline README.md:10]: Please, leave the version bumps to another commit.

File: internal/builder/main.go [hunk 2]
```diff
@@ -83,15 +88,10 @@ func Generate(cfg Config) error {
...
-	goBinary := cfg.Distribution.Go
 	// first, we test to check if we have Go at all
-	if _, err := exec.Command(goBinary, "env").CombinedOutput(); err != nil {
-		path, err := exec.LookPath("go")
-		if err != nil {
-			return ErrGoNotFound
-		}
-		goBinary = path
-		cfg.Logger.Info("Using go from PATH", "Go executable", path)
+	goBinary, err := GetGoPath(cfg)
```
Comment (jpkrohling) [inline internal/builder/main.go:92]: This could be private to the package.

File: internal/builder/main.go [hunk 3]
```diff
@@ -105,6 +105,39 @@ func Compile(cfg Config) error {
...
+	if err != nil {
+		return err
+	}
+
+	cfg.Logger.Info("Getting go modules")
+	cmd := exec.Command(goBinary, "mod", "download")
+	cmd.Dir = cfg.Distribution.OutputPath
+	if out, err := cmd.CombinedOutput(); err != nil {
+		return fmt.Errorf("failed to get go modules: %w. Output: %q", err, out)
```
Comment (jpkrohling) [inline internal/builder/main.go:120]:
```suggestion
		return fmt.Errorf("failed to download go modules: %w. Output: %q", err, out)
```

Comment (jpkrohling) [reply internal/builder/main.go:120]: Is it possible to determine whether the failure was permanent? Does it make sense to retry, especially on network connectivity issues?

File: internal/builder/main.go [hunk 4]
```diff
@@ -105,6 +105,39 @@ func Compile(cfg Config) error {
 	return nil
 }
 
+// GetModules Runs go mod download to update go.mod and go.sum
```
Comment (jpkrohling) [inline internal/builder/main.go:108]:
```suggestion
// GetModules retrieves the go modules, updating go.mod and go.sum in the process
```

File: internal/builder/main.go [hunk 5]
```diff
@@ -105,6 +105,39 @@ func Compile(cfg Config) error {
...
+	cmd := exec.Command(goBinary, "mod", "download")
+	cmd.Dir = cfg.Distribution.OutputPath
+	if out, err := cmd.CombinedOutput(); err != nil {
+		return fmt.Errorf("failed to get go modules: %w. Output: %q", err, out)
+	}
+
+	return nil
+}
+
+// GetGoPath Checks if go is present and correct and returns a useable go bin location
```
Comment (jpkrohling) [inline internal/builder/main.go:126]:
```suggestion
// GetGoPath checks if go is present and correct, and returns a useable go bin location
```

**Review by jpkrohling — COMMENTED** [review]
Looks great, but please revert the version bumps, as they should go as part of their commit (like #19)

File: README.md (hunk 1 above)
Comment (eroteme) [reply README.md:10]: Yes, I thought I'd branched cleanly .. obviously not. I'll revert the version bump changes from this pull request

File: internal/builder/main.go (hunk 3 above)
Comment (eroteme) [reply internal/builder/main.go:120]: I'll look if `go mod download` gives information to the type of the error (network, transient etc) and how easy it can be parsed. If go mod gives this information I could then retry but wonder how many times/ how long it should retry for. For example we run this in a build pipeline which has a timeout. If it took too long to retry it would cause the pipeline to timeout.

**Comment by eroteme** [comment]
I've reverted the version changes from this pull request but hope it doesn't revert them in main when merged. Not sure why by branch had these changes in as I branched from a previous commit

**Comment by jpkrohling** [comment]
It will revert the ones from the main branch :-)

File: internal/builder/main.go (hunk 3 above)
Comment (jpkrohling) [reply internal/builder/main.go]: To make it easier to get this fix ready for 1.16, could you create an issue tracking this idea? Or perhaps even simpler: try at most 3 times before failing permanently.

Comment (eroteme) [reply internal/builder/main.go]: Added basic retry. Will create an issue to improve at a later stage

File: README.md [hunk 6]
```diff
@@ -7,7 +7,7 @@ This program generates a custom OpenTelemetry Collector binary based on a given
 $ GO111MODULE=on go get github.com/open-telemetry/opentelemetry-collector-builder
 $ cat > ~/.otelcol-builder.yaml <<EOF
 exporters:
-  - gomod: "github.com/open-telemetry/opentelemetry-collector-contrib/exporter/alibabacloudlogserviceexporter v0.22.0"
+  - gomod: "github.com/open-telemetry/opentelemetry-collector-contrib/exporter/alibabacloudlogserviceexporter v0.21.0"
```
Comment (jpkrohling) [inline README.md:10]: These changes should be reverted. Would you like me to revert it for you?

Comment (eroteme) [reply README.md:10]: Yes I thought I'd removed those. could be quicker if you revert. Thanks

**Review by jpkrohling — APPROVED** [review]
